### PR TITLE
[codegen] Emit missing cleanups for stmt-expr and coro suspensions [take-2]

### LINKED ISSUE #63818: destructors not always properly run when a coroutine is suspended and then destroyed
[Example](https://godbolt.org/z/zezvTqY8v):

```c++
#include <coroutine>
#include <iostream>

struct coroutine {
  struct promise_type;
  std::coroutine_handle<promise_type> handle;
  ~coroutine() { handle.destroy(); }
};

struct coroutine::promise_type {
  coroutine get_return_object() {
    return {std::coroutine_handle<promise_type>::from_promise(*this)};
  }
  std::suspend_never initial_suspend() noexcept { return {}; }
  std::suspend_never final_suspend() noexcept { return {}; }
  void return_void() {}
  void unhandled_exception() {}
};

struct Printy {
  Printy(const char *name) : name(name) { std::cout << "Printy(" << name << ")\n"; }
  Printy(const Printy&) = delete;
  ~Printy() { std::cout << "~Printy(" << name << ")\n"; }
  const char *name;
};

int main() {
  [] -> coroutine {
    Printy a("a");
    Printy arr[] = {
      Printy("b"), Printy("c"),
      (co_await std::suspend_always{}, Printy("d")),
      Printy("e")
    };
  }();
}
```

When the coroutine is destroyed after being suspended, `a` is destroyed, but `arr[0]` and `arr[1]` are not. Clang does not in general properly create cleanups for non-exceptional control flow that occurs in the middle of an expression / an initializer. At least array initialization is missing cleanups here, but it'd be worth checking through all exception-only cleanups because most of them are probably incorrect. It looks like there are 15 places where we currently push an EH-only cleanup:

```
CGCall.cpp:      pushFullExprCleanup<DestroyUnpassedArg>(EHCleanup, Slot.getAddress(),
CGClass.cpp:    CGF.EHStack.pushCleanup<CallBaseDtor>(EHCleanup, BaseClassDecl,
CGClass.cpp:    EHStack.pushCleanup<CallDelegatingCtorDtor>(EHCleanup,
CGCoroutine.cpp:    EHStack.pushCleanup<CallCoroEnd>(EHCleanup);
CGDecl.cpp:  pushDestroy(EHCleanup, addr, type, getDestroyer(dtorKind), true);
CGDecl.cpp:  pushFullExprCleanup<IrregularPartialArrayDestroy>(EHCleanup,
CGDecl.cpp:  pushFullExprCleanup<RegularPartialArrayDestroy>(EHCleanup,
CGException.cpp:  pushFullExprCleanup<FreeException>(EHCleanup, addr.getPointer());
CGExprAgg.cpp:        CGF.pushDestroy(EHCleanup, LV.getAddress(CGF), CurField->getType(),
CGExprAgg.cpp:        CGF.pushDestroy(EHCleanup, LV.getAddress(CGF), field->getType(),
CGExprCXX.cpp:      .pushCleanupWithExtra<DirectCleanup>(EHCleanup,
CGExprCXX.cpp:    .pushCleanupWithExtra<ConditionalCleanup>(EHCleanup,
ItaniumCXXABI.cpp:    CGF.EHStack.pushCleanup<CallGuardAbort>(EHCleanup, guard);
MicrosoftCXXABI.cpp:    CGF.EHStack.pushCleanup<ResetGuardBit>(EHCleanup, GuardAddr, GuardNum);
MicrosoftCXXABI.cpp:    CGF.EHStack.pushCleanup<CallInitThreadAbort>(EHCleanup, GuardAddr);
```

This bug is not new with coroutines; the same thing happens with statement expressions:

```c++
int main() {
  Printy arr[] = { Printy("a"), ({ return 0; Printy("b"); }) };
}
```

... never destroys `arr[0]`. But it seems more pressing now that it's reachable from standard C++20 code.

### PR DESCRIPTION
**Pull request by usx95**
Fixes https://github.com/llvm/llvm-project/issues/63818 for control flow out of an expressions.

#### Background

A control flow could happen in the middle of an expression due to stmt-expr and coroutine suspensions.

Due to branch-in-expr, we miss running cleanups for the temporaries constructed in the expression before the branch.
Previously, these cleanups were only added as `EHCleanup` during the expression and as normal expression after the full expression.

Examples of such deferred cleanups include:

`ParenList/InitList`: Cleanups for fields are performed by the destructor of the object being constructed.
`Array init`: Cleanup for elements of an array is included in the array cleanup.
`Lifetime-extended temporaries`: reference-binding temporaries in braced-init are lifetime extended to the parent scope.
`Lambda capture init`: init in the lambda capture list is destroyed by the lambda object.

---

#### In this PR

In this PR, we change some of the `EHCleanups` cleanups to `NormalAndEHCleanups` to make sure these are emitted when we see a branch inside an expression (through statement expressions or coroutine suspensions).

These are supposed to be deactivated after full expression and destroyed later as part of the destructor of the aggregate or array being constructed. To simplify deactivating cleanups, we add two utilities as well:
* `DeferredDeactivationCleanupStack`: A stack to remember cleanups with deferred deactivation.
* `CleanupDeactivationScope`: RAII for deactivating cleanups added to the above stack.

---

#### Deactivating normal cleanups
These were previously `EHCleanups` and not `Normal` and **deactivation** of **required** `Normal` cleanups had some bugs. These specifically include deactivating `Normal` cleanups which are not the top of `EHStack` [source1](https://github.com/llvm/llvm-project/blob/92b56011e6b61e7dc1628c0431ece432f282b3cb/clang/lib/CodeGen/CGCleanup.cpp#L1319), [2](https://github.com/llvm/llvm-project/blob/92b56011e6b61e7dc1628c0431ece432f282b3cb/clang/lib/CodeGen/CGCleanup.cpp#L722-L746). This has not been part of our test suite (maybe it was never required before statement expressions). In this PR, we also fix the emission of required-deactivated-normal cleanups.

---

#### Test ideas
I have tried out printf-style ctor-dtor calls in the following cases manually and verified that the correct dtors are invoked: https://godbolt.org/z/jMYTK7Kxe
I would be open to ideas for better tests than just lit-test for LLVM IR. We might benefit from some integration tests which can run the program and assert

### CODE REVIEW

**Comment by usx95** [comment]
Friendly ping for review: @jyknight @zygoloid @efriedma-quic

**Comment by Sirraide** [comment]
I’m the wrong person to ping if it’s anything codegen-related; I’ll leave reviewing this to people more familiar w/ that part of Clang.

File: clang/lib/AST/Expr.cpp [hunk 1]
```diff
@@ -2503,6 +2505,26 @@ Stmt *BlockExpr::getBody() {
...
+    bool HasBranch = false;
+    bool activate() {
+      HasBranch = true;
+      return false;
+    }
+    // Coroutine suspensions.
+    bool VisitCoawaitExpr(CoawaitExpr *) { return activate(); }
+    bool VisitCoyieldExpr(CoyieldExpr *) { return activate(); }
+    // Control flow in stmt-expressions.
+    bool VisitBreakStmt(BreakStmt *) { return activate(); }
```
Comment (AaronBallman) [inline clang/lib/AST/Expr.cpp:2519]: Should this also handle continue statements?

File: clang/lib/AST/Expr.cpp [hunk 2]
```diff
@@ -2503,6 +2505,26 @@ Stmt *BlockExpr::getBody() {
...
+bool Expr::mayBranchOut() const {
+  struct BranchDetector : public RecursiveASTVisitor<BranchDetector> {
+    bool HasBranch = false;
+    bool activate() {
+      HasBranch = true;
+      return false;
+    }
+    // Coroutine suspensions.
+    bool VisitCoawaitExpr(CoawaitExpr *) { return activate(); }
+    bool VisitCoyieldExpr(CoyieldExpr *) { return activate(); }
```
Comment (AaronBallman) [inline clang/lib/AST/Expr.cpp:2517]: Should this also handle `co_return` statements?

File: clang/lib/AST/Expr.cpp [hunk 3]
```diff
@@ -2503,6 +2505,26 @@ Stmt *BlockExpr::getBody() {
...
+    // Control flow in stmt-expressions.
+    bool VisitBreakStmt(BreakStmt *) { return activate(); }
+    bool VisitReturnStmt(ReturnStmt *) { return activate(); }
+    bool VisitGotoStmt(GotoStmt *) { return activate(); }
```
Comment (AaronBallman) [inline clang/lib/AST/Expr.cpp:2521]: Should this also handle indirect goto statements?

File: clang/lib/AST/Expr.cpp [hunk 4]
```diff
@@ -2503,6 +2505,26 @@ Stmt *BlockExpr::getBody() {
...
+  };
```
Comment (AaronBallman) [inline clang/lib/AST/Expr.cpp:2522]: Should we consider `throw` to be a kind of branching control flow as well?

**Review by AaronBallman — COMMENTED** [review]
Hopefully @rjmccall or @efriedma-quic can take a look at the codegen bits. I did have some questions about other kinds of flow control statements we might want to cover.

File: clang/lib/AST/Expr.cpp (hunk 3 above)
Comment (usx95) [reply clang/lib/AST/Expr.cpp:2521]: Added.

File: clang/lib/AST/Expr.cpp (hunk 4 above)
Comment (usx95) [reply clang/lib/AST/Expr.cpp:2522]: I think it should be fine to stick to non-exceptional control flow out of this expression. `throw` or, for that matter, any call to a throwing function would be branching but that is already dealt with exceptions.

File: clang/lib/AST/Expr.cpp (hunk 1 above)
Comment (usx95) [reply clang/lib/AST/Expr.cpp:2519]: Yes. Added a test case for it as well.

**Review by AaronBallman — COMMENTED** [review]
I didn't look too hard at the codegen bits, but the Expr bits look good to me. Perhaps we want this test to live in CodeGen rather than CodeGenCoroutines given that it covers so many non-coroutine test cases as well?

**Comment by usx95** [comment]
> Perhaps we want this test to live in CodeGen rather than CodeGenCoroutines given that it covers so many non-coroutine test cases as well?
Makes sense. I split the tests into coroutine and stmt-expr (didn't want to depend on "Inputs/coroutine.h" from CodegenCXX).

File: clang/lib/AST/Expr.cpp [hunk 5]
```diff
@@ -2503,6 +2506,29 @@ Stmt *BlockExpr::getBody() {
 // Generic Expression Routines
 //===----------------------------------------------------------------------===//
 
+bool Expr::mayBranchOut() const {
```
Comment (zygoloid) [inline clang/lib/AST/Expr.cpp:2509]: Computing this seems a little expensive in general. I wonder if we could track a bit on `FunctionDecl` that indicates whether it contains any branches out of statement expressions, and skip calling this if the enclosing function is not a coroutine and does not contain branches out of statement expressions.

**Comment by efriedma-quic** [comment]
Instead of Expr::mayBranchOut, I'd prefer to just unconditionally create the alloca, then delete it later if it turns out we didn't actually need to emit the branch.  Trying to explicitly compute whether there's a branch out seems both difficult, and potentially costly for compile-time.

I like the unified approach here for eh and non-eh cleanups.

**Comment by efriedma-quic** [comment]
If we do keep mayBranchOut(), "asm goto" should also be added to the list of expressions that can branch out.  (I think JumpDiagnostics currently forbids actually branching out in cases that would require a cleanup, but better to be defensive here.)

**Comment by usx95** [comment]
> Instead of Expr::mayBranchOut, I'd prefer to just unconditionally create the alloca, then delete it later if it turns out we didn't actually need to emit the branch. 

I had earlier tried tracking instructions auxiliary to a particular cleanup in #83224 ([src](https://github.com/llvm/llvm-project/pull/83224/files#diff-9cdaea6a793ed2892bfcd6b431e933a49ebb25caa2bd1d630cd1ca823281092aR263-R286)). This gets ugly very quickly and adds quite some complexity to the cleanup addition and emission code. For example, more instructions could be added if the cleanup is conditional.

> Trying to explicitly compute whether there's a branch out seems both difficult, and potentially costly for compile-time.

Computing this 100% accurately is indeed difficult. But if we allow false-positives, as in the current approach, it gets clearer and simpler. As far as compile-time is concerned, this involves revisiting the complete expression exactly one more time. We could, in principle, cache this per expression to keep this linear.

**Comment by efriedma-quic** [comment]
> > Instead of Expr::mayBranchOut, I'd prefer to just unconditionally create the alloca, then delete it later if it turns out we didn't actually need to emit the branch.
> 
> I had earlier tried tracking instructions auxiliary to a particular cleanup in #83224 ([src](https://github.com/llvm/llvm-project/pull/83224/files#diff-9cdaea6a793ed2892bfcd6b431e933a49ebb25caa2bd1d630cd1ca823281092aR263-R286)). This gets ugly very quickly and adds quite some complexity to the cleanup addition and emission code. For example, more instructions could be added if the cleanup is conditional.

You could simplify that code significantly, I think: instead of adding every instruction to the list, just add the alloca, then recursively delete all instructions that use the alloca.

> As far as compile-time is concerned, this involves revisiting the complete expression exactly one more time.

You can end up visiting more than once if an array init contains another array init.

File: clang/lib/AST/Expr.cpp (hunk 5 above)
Comment (usx95) [reply clang/lib/AST/Expr.cpp]: This is now no more needed. We switched to deleting instructions related to unemitted cleanups.

**Comment by usx95** [comment]
Thanks @efriedma-quic . I was not aware that we have such edges in Codegen which could help in deleting these instructions transitively. Switched to capturing only auxillary `allocas` and deleting all transitive uses. PTAL.

File: clang/lib/CodeGen/CGExprCXX.cpp [hunk 6]
```diff
@@ -1105,19 +1105,24 @@ void CodeGenFunction::EmitNewArrayInitializer(
     }
 
     // Enter a partial-destruction Cleanup if necessary.
-    if (needsEHCleanup(DtorKind)) {
+    if (DtorKind) {
+      AllocaTrackerRAII AllocaTracker(*this);
```
[the remark this answers is not in the document]
Comment (usx95) [reply clang/lib/CodeGen/CGExprCXX.cpp:1109]: Unfortunately, that is not the case. We do create more allocas in conditional branches in `pushFullExprCleanup`, for eg: create `cond-cleanup.save` allocas in `DominatingLLVMValue::save` ([src](https://github.com/llvm/llvm-project/blob/e93b5f5a4776ffea12d03652559dfdf8d421184c/clang/lib/CodeGen/CodeGenFunction.h#L5184)).

[the remark this answers is not in the document]
Comment (usx95) [reply clang/lib/CodeGen/CGExprCXX.cpp:1109]: Bump. @efriedma-quic does this look reasonable ?

**Review by efriedma-quic — APPROVED** [review]
LGTM

**Comment by ayermolo** [comment]
I think this commit is causing build failure when building clangd in debug mode with clang built in release mode.

```
Instruction does not dominate all uses!
  %K1104 = getelementptr inbounds %"struct.llvm::json::Object::KV", ptr %arrayinit.begin1100, i32 0, i32 0, !dbg !93928
  call void @_ZN4llvm4json9ObjectKeyD2Ev(ptr noundef nonnull align 8 dereferenceable(24) %K1104) #21, !dbg !93937
fatal error: error in backend: Broken module found, compilation aborted!
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: /home/ayermolo/local/llvm-build-upstream-release/bin/clang++ -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Itools/clang/tools/extra/clangd -I/home/ayermolo/local/upstream-llvm/llvm-project/clang-tools-extra/clangd -I/home/ayermolo/local/upstream-llvm/llvm-project/clang-tools-extra/clangd/../include-cleaner/include -Itools/clang/tools/extra/clangd/../clang-tidy -I/home/ayermolo/local/upstream-llvm/llvm-project/clang/include -Itools/clang/include -Iinclude -I/home/ayermolo/local/upstream-llvm/llvm-project/llvm/include -I/home/ayermolo/local/upstream-llvm/llvm-project/clang-tools-extra/pseudo/lib/../include -isystem/home/ayermolo/local/llvm-build-upstream-llvm/build/lib/clang/19/include -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -fno-common -Woverloaded-virtual -Wno-nested-anon-types -g -std=c++17 -MD -MT /home/ayermolo/local/llvm-build-upstream-debug/tools/clang/tools/extra/clangd/CMakeFiles/obj.clangDaemon.dir/ClangdLSPServer.cpp.o -MF /home/ayermolo/local/llvm-build-upstream-debug/tools/clang/tools/extra/clangd/CMakeFiles/obj.clangDaemon.dir/ClangdLSPServer.cpp.o.d -o /home/ayermolo/local/llvm-build-upstream-debug/tools/clang/tools/extra/clangd/CMakeFiles/obj.clangDaemon.dir/ClangdLSPServer.cpp.o -c /home/ayermolo/local/upstream-llvm/llvm-project/clang-tools-extra/clangd/ClangdLSPServer.cpp
```

Dropping this commit and running above clang build command makes it pass.
build command for easier reading:
```
COMP=/home/ayermolo/local/llvm-build-upstream-release/bin
$COMP/clang++ -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS \
  -Itools/clang/tools/extra/clangd \
  -I/home/ayermolo/local/upstream-llvm/llvm-project/clang-tools-extra/clangd \
  -I/home/ayermolo/local/upstream-llvm/llvm-project/clang-tools-extra/clangd/../include-cleaner/include \
  -Itools/clang/tools/extra/clangd/../clang-tidy \
  -I/home/ayermolo/local/upstream-llvm/llvm-project/clang/include \
  -Itools/clang/include \
  -Iinclude \
  -I/home/ayermolo/local/upstream-llvm/llvm-project/llvm/include \
  -I/home/ayermolo/local/upstream-llvm/llvm-project/clang-tools-extra/pseudo/lib/../include \
  -isystem/home/ayermolo/local/llvm-build-upstream-llvm/build/lib/clang/19/include \
  -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra \
  -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi \
  -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override \
  -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -fno-common -Woverloaded-virtual \
  -Wno-nested-anon-types -g -std=c++17 -MD \
  -MT /home/ayermolo/local/llvm-build-upstream-debug/tools/clang/tools/extra/clangd/CMakeFiles/obj.clangDaemon.dir/ClangdLSPServer.cpp.o \
  -MF /home/ayermolo/local/llvm-build-upstream-debug/tools/clang/tools/extra/clangd/CMakeFiles/obj.clangDaemon.dir/ClangdLSPServer.cpp.o.d \
  -o  /home/ayermolo/local/llvm-build-upstream-debug/tools/clang/tools/extra/clangd/CMakeFiles/obj.clangDaemon.dir/ClangdLSPServer.cpp.o \
  -c  /home/ayermolo/local/upstream-llvm/llvm-project/clang-tools-extra/clangd/ClangdLSPServer.cpp
```

**Comment by ayermolo** [comment]
@usx95 can you repro?
Also is there ETA on a fix, and if not can you revert this?

**Comment by usx95** [comment]
[ayermolo](https://github.com/ayermolo) Thanks for the reproducer. I was able to reproduce it. Sent out fix https://github.com/llvm/llvm-project/pull/88670

**Pull request merged**